cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,227 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package upgrader
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/option"
+)
+
+// Parameter key shared with external packages
+const (
+	ClientIDFlagKey = "client_id"
+	DefaultTimeout  = "90m"
+)
+
+const (
+	logPrefix = "[windows-upgrade]"
+
+	metadataKeyWindowsStartupScriptURL       = "windows-startup-script-url"
+	metadataKeyWindowsStartupScriptURLBackup = "windows-startup-script-url-backup"
+
+	versionWindows2008r2 = "windows-2008r2"
+	versionWindows2012r2 = "windows-2012r2"
+)
+
+var (
+	supportedSourceOSVersions = map[string]string{versionWindows2008r2: versionWindows2012r2}
+	supportedTargetOSVersions = reverseMap(supportedSourceOSVersions)
+
+	upgradeScriptName = map[string]string{versionWindows2008r2: "upgrade_script_2008r2_to_2012r2.ps1"}
+
+	expectedCurrentLicense = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2008-r2-dc"}
+	licenseToAdd           = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2012-r2-dc-in-place-upgrade"}
+)
+
+type derivedVars struct {
+	project string
+	zone    string
+
+	osDiskURI        string
+	osDiskType       string
+	osDiskDeviceName string
+	osDiskAutoDelete bool
+
+	instanceName           string
+	machineImageBackupName string
+	osDiskSnapshotName     string
+	newOSDiskName          string
+	installMediaDiskName   string
+
+	windowsStartupScriptURLBackup       *string
+	windowsStartupScriptURLBackupExists bool
+}
+
+// Upgrader implements upgrading logic.
+type Upgrader struct {
+	// Input params
+	ClientID               string
+	InstanceURI            string
+	SkipMachineImageBackup bool
+	AutoRollback           bool
+	SourceOS               string
+	TargetOS               string
+	ProjectPtr             *string
+	Timeout                string
+	ScratchBucketGcsPath   string
+	Oauth                  string
+	Ce                     string
+	GcsLogsDisabled        bool
+	CloudLogsDisabled      bool
+	StdoutLogsDisabled     bool
+	CurrentExecutablePath  string
+
+	*derivedVars
+
+	ctx context.Context
+}
+
+type upgraderInterface interface {
+	getUpgrader() *Upgrader
```

Comment:
You've got a circular dependency here -- what's the motivation for exposing your implementation?

Reply:
Upgrader is a lib so that it can be referenced by not only this specific main function.
As a lib, we more or less need to expose something. I exposed Upgrader because it contains input params, which should be populated by main func. Since now Upgrader contains not only data but also implementations of methods, it looks like exposed too much. So in the latest commit I wrap these input params as a struct UpgraderInputParams, and only expose it. Both upgrader and its interface are hidden now.

The only purpose of the interface is for testability.